BootloaderCommonPkg/Library/ElfLib/ElfLib.c

```diff
@@ -0,0 +1,228 @@
+  UINT32                      Msize;
+  UINT32                      Fsize;
+  UINT32                      Mpos;
+  UINT32                      Ipos;
+
+  //
+  // TODO: Check if image, load region overlap; if yes, move image away.
+  //
+
+  //
+  // Get ELF header and Program header
+  //
+  Ehdr = (ELF32_EHDR *) ImageBase;
+  Phdr = (ELF32_PHDR *) ((UINT32)ImageBase + Ehdr->E_Phoff);
+
+  //
+  //  Iterate over the program headers and load all necessary segments.
+  //  Keep track of the highest load address used in <eaddr>
+  //
+  Eaddr = 0;
+  for (Index = 0; Index < Ehdr->E_Phnum ; Index ++) {
+    if (Phdr[Index].P_Type != PT_LOAD) {
+      continue;
+    }
+
+    Msize = Phdr[Index].P_Memsz;            // target (in-memory) segment size
+    Fsize = Phdr[Index].P_Filesz;           // source (image) segment size
+    Mpos  = Phdr[Index].P_Paddr;            // target segment address
+    Ipos  = Phdr[Index].P_Offset;           // location of segment in image
+
+    if ((Mpos <= (UINT32)ImageBase) && (Mpos + Msize > (UINT32)ImageBase)) {
+      DEBUG ((DEBUG_INFO, "** IMAGE overwrite, out of range **\n"));
+      return NULL;
+    }
+
+    if (Fsize != 0) {
+      CopyMem ((VOID *) Mpos, (VOID *) ((UINT32)ImageBase + Ipos), Fsize);
+    }
+
+    if (Msize > Fsize) {
+      ZeroMem ((VOID *) (Mpos + Fsize), Msize - Fsize);
+    }
+
+    Eaddr = MAX (Eaddr, Mpos + Msize);
+  }
+  ASSERT (Eaddr != 0);
+
+  //
+  // The image is now bootable; try to load the symbol table as well
+  //
+  LoadElfSymtab (ImageBase, ALIGN_UP (Eaddr, SIZE_4KB));
+
+  //
+  // Return Entry Point
+  //
+  return (UINT32 *) Ehdr->E_Entry;
+}
+
+/**
+  Update MultibootInfo from ELF
+
+  This function updates Multiboot Information if there is symbol information in ELF.
+  Caller must guarantee the image is ELF, multiboot and loaded successfully.
+
+  @param[in]        ImageBase   Memory address of an image
+  @param[in, out]   MbInfo      Pointer to MULTIBOOT_INFO which needs to be updated
+
+  @retval RETURN_SUCCESS        MULTIBOOT_INFO successfully updated.
+  @retval RETURN_NOT_FOUND      No ELF Symbol information found.
+
+**/
+RETURN_STATUS
+UpdateElfMultibootInfo (
+  IN      UINT32          ImageBase,
+  IN  OUT MULTIBOOT_INFO  *MbInfo
+  )
+{
+  ELF32_EHDR                 *Ehdr;
+  ELF32_PHDR                 *Phdr;
+  UINT32                      Eaddr;
+  UINT32                      Index;
+
+  ASSERT (MbInfo != NULL);
+  MbInfo->Flags &= ~(MULTIBOOT_INFO_HAS_AOUT_SYMS | MULTIBOOT_INFO_HAS_ELF_SYMS);
+
+  Ehdr = (ELF32_EHDR *) (UINTN) ImageBase;
+  Phdr = (ELF32_PHDR *) (UINTN) (ImageBase + Ehdr->E_Phoff);
+
+  if (Ehdr->E_Shnum == 0 || Ehdr->E_Shentsize) {
+    return RETURN_NOT_FOUND;
+  }
+
+  Eaddr = 0;
+  for (Index = 0; Index < Ehdr->E_Phnum ; Index ++) {
+    if (Phdr[Index].P_Type != PT_LOAD) {
+      continue;
+    }
+    Eaddr = MAX (Eaddr, Phdr[Index].P_Paddr + Phdr[Index].P_Memsz);
+  }
+
+  MbInfo->Flags       |= MULTIBOOT_INFO_HAS_ELF_SYMS;
```

Comment:
I think the `MULTIBOOT_INFO_HAS_ELF_SYMS` name is a bit misleading and does not exactly line up with [the spec](https://www.gnu.org/software/grub/manual/multiboot/multiboot.html) which defines this bit as:
```
     /* is there an ELF section header table? */
     #define MULTIBOOT_INFO_ELF_SHDR                 0X00000020
```

Reply:
Yes, it seems. Let me include the change in next commit rather than new PR.

Reply:
I double checked it. It may need another PR since other bits need to be updated as well. Let me use the name as it is in this commit.